include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+    std::lock_guard<std::mutex> lock(mutex_);
+    while (!pending_.empty()) {
+      collect_pending_unsafe(ctx);
+      if (pending_.empty()) {
+        break;
+      }
+      std::this_thread::yield();
+    }
+  }
+
+  inline std::unique_ptr<alloc_type, std::function<void(alloc_type*)>>
+  get_unique(cudaStream_t ctx = nullptr) {
+    alloc_type* ptr;
+    get_raw(&ptr, (&ptr) + 1, ctx);
+    return {ptr, [this, ctx](alloc_type* p) { put_raw(&p, (&p) + 1, ctx); }};
+  }
+
+  inline std::shared_ptr<alloc_type> get_shared(cudaStream_t ctx = nullptr) {
+    alloc_type* ptr;
+    get_raw(&ptr, (&ptr) + 1, ctx);
+    return {ptr, [this, ctx](alloc_type* p) { put_raw(&p, (&p) + 1, ctx); }};
+  }
+
+  template <size_t N>
+  inline StaticWorkspace<N> get_workspace(cudaStream_t ctx = nullptr) {
+    return {this, ctx};
+  }
+
+  inline DynamicWorkspace get_workspace(size_t n, cudaStream_t ctx = nullptr) {
+    return {this, n, ctx};
+  }
+
+  friend std::ostream& operator<<<Allocator>(std::ostream&, const MemoryPool&);
+
+ private:
+  inline void collect_pending_unsafe(cudaStream_t ctx) {
+    auto it = std::remove_if(
+        pending_.begin(), pending_.end(),
+        [this, ctx](const std::pair<alloc_type*, cudaEvent_t>& pair) {
+          const cudaError_t event_state = cudaEventQuery(pair.second);
+          switch (event_state) {
+            case cudaSuccess:
+              // Stock buffers and destroy those that
+              // are no longer needed.
+              if (stock_.size() < options.max_stock) {
+                stock_.emplace_back(pair.first);
+              } else {
+                Allocator::free(pair.first, ctx);
+              }
+              ready_events_.emplace_back(pair.second);
+              return true;
+            case cudaErrorNotReady:
+              return false;
+            default:
+              CUDA_CHECK(event_state);
+              return false;
+          }
+        });
+    pending_.erase(it, pending_.end());
+  }
+
+  template <class Iterator>
+  inline void get_raw(Iterator first, Iterator const last, cudaStream_t ctx) {
+    // Get pre-allocated buffers if stock available.
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+
+      for (; first != last; ++first) {
+        // If no buffers available, try to make some available.
+        if (stock_.empty()) {
+          collect_pending_unsafe(ctx);
+          if (stock_.empty()) {
+            // No buffers available.
+            break;
+          }
+        }
+
+        // Just take the next available buffer.
+        *first = stock_.back();
+        stock_.pop_back();
+      }
+    }
+
+    // Forge new buffers until request can be filled.
+    for (; first != last; ++first) {
+      *first = Allocator::alloc(options.buffer_size, ctx);
+    }
+  }
+
+  template <class Iterator>
+  inline void put_raw(Iterator first, Iterator const last, cudaStream_t ctx) {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    if (ctx) {
```

Comment:
`cudaStream_t stream = 0`

Reply:
This is an internally used function. Caller should always provider a ctx/stream. Although, it could be 0. No need for default argument.

Reply:
P.S.: Just in case you didn't notice. This is a `private` method.